File: skills/code-architect/references/code-examples/go/facade.go
// Facade: expose a single simplified entry point that hides the wiring of
// a more complex subsystem.
package main

import "fmt"

// Subsystem pieces, each with its own narrow responsibility.
type loader struct{}

func (loader) load() string { return "raw" }

type parser struct{}

func (parser) parse(in string) string { return in + "-parsed" }

type validator struct{}

func (validator) validate(in string) bool { return in != "" }

// Facade orchestrates the subsystem behind one method.
type Pipeline struct {
	l loader
	p parser
	v validator
}

func (pl Pipeline) Run() (string, bool) {
	data := pl.p.parse(pl.l.load())
	return data, pl.v.validate(data)
}

func main() {
	out, ok := Pipeline{}.Run()
	fmt.Println(out, ok)
}

File: skills/code-architect/references/code-examples/typescript/Facade.ts
// Facade — a simplified interface over a complex subsystem.

class VideoDecoder {
  decode(file: string): string {
    return `frames(${file})`;
  }
}

class AudioMixer {
  mix(file: string): string {
    return `audio(${file})`;
  }
}

class SubtitleLoader {
  load(file: string): string {
    return `subs(${file})`;
  }
}

// Facade hides subsystem wiring behind one method.
class MediaPlayerFacade {
  private decoder = new VideoDecoder();
  private mixer = new AudioMixer();
  private subtitles = new SubtitleLoader();

  play(file: string): string {
    const video = this.decoder.decode(file);
    const audio = this.mixer.mix(file);
    const subs = this.subtitles.load(file);
    return `playing -> ${video} + ${audio} + ${subs}`;
  }
}

function demo(): void {
  const player = new MediaPlayerFacade();
  console.log(player.play("movie.mkv"));
}

demo();
